File: 098_Implement Trie (Prefix Tree).py
"""
Implement Trie (Prefix Tree)
Insert, search and startsWith methods.
"""
from typing import Dict

class TrieNode:
    def __init__(self):
        self.children: Dict[str, 'TrieNode'] = {}
        self.end = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    def insert(self, word: str) -> None:
        p = self.root
        for c in word:
            if c not in p.children: p.children[c] = TrieNode()
            p = p.children[c]
        p.end = True
    def search(self, word: str) -> bool:
        p = self.root
        for c in word:
            if c not in p.children: return False
            p = p.children[c]
        return p.end
    def startsWith(self, prefix: str) -> bool:
        p = self.root
        for c in prefix:
            if c not in p.children: return False
            p = p.children[c]
        return True

if __name__ == '__main__':
    t = Trie(); t.insert('apple'); print(t.search('apple'), t.search('app'), t.startsWith('app'))
    t.insert('app'); print(t.search('app'))

# Time Complexity: O(L)
# Space Complexity: O(total chars)

File: 098_Implement Trie (Prefix Tree).cpp
// Implement Trie (Prefix Tree)
// Insert, search and startsWith operations.
#include <bits/stdc++.h>
using namespace std;

struct TrieNode {
    bool end=false; array<TrieNode*,26> ch{}; TrieNode(){ ch.fill(nullptr); }
};

class Trie {
    TrieNode* root;
public:
    Trie(): root(new TrieNode()){}
    void insert(const string& w){ TrieNode* p=root; for(char c:w){ int i=c-'a'; if(!p->ch[i]) p->ch[i]=new TrieNode(); p=p->ch[i]; } p->end=true; }
    bool search(const string& w){ TrieNode* p=root; for(char c:w){ int i=c-'a'; if(!p->ch[i]) return false; p=p->ch[i]; } return p->end; }
    bool startsWith(const string& pfx){ TrieNode* p=root; for(char c:pfx){ int i=c-'a'; if(!p->ch[i]) return false; p=p->ch[i]; } return true; }
};

int main(){
    Trie t; t.insert("apple"); cout<<boolalpha<<t.search("apple")<<" "<<t.search("app")<<" "<<t.startsWith("app")<<"\n";
    t.insert("app"); cout<<t.search("app")<<"\n";
    return 0;
}

// Time Complexity: insert/search/prefix O(L)
// Space Complexity: O(total chars)
